src/main/java/org/owasp/esapi/reference/DefaultEncoder.java

```diff
@@ -305,13 +305,18 @@ public String encodeForLDAP(String input, boolean encodeWildcards) {
         }
         // TODO: replace with LDAP codec
         StringBuilder sb = new StringBuilder();
+        // According to "Special Characters" at [1], the encoder should escape '*', '(', ')', '\', '/', NUL. Also see [2].
```

Comment:
@noloader - In reference [1], that Microsoft documentation definitely lists '/' as requiring encoding, but that reference also references [RFC 2254](https://datatracker.ietf.org/doc/html/rfc2254), and if you look at Section 4 of that document (which defines search filters for LDAPv3), you will not find '/' as being mentioned as requiring encoding.  So, I am thinking this is something that is Active Directory specific. (Thanks Microsoft! Another standard ignored.)

Consequently, I'm not really sure what to do about this. It seems important in that AD is one of the most common LDAP interfaces that people deal with, but if we add we make encoding '/' the default, which I'm not really for, it could break other pure LDAPv3 search filters that conform specifically to RFC 2254.

I guess one ugly option would be to define and document an new `Encoder` interface method that looks like:
```
    public String encodeForLDAP(String input, boolean encodeWildcards, boolean isActiveDirectory)
```
and have that last argument be treated by the other 2 `encodeForLDAP()` methods as the `isActiveDirectory` flag being set to `false`.

That's not perfect, but at least it would keep from breaking things with non-AD LDAPv3 compliant directory servers. The only other (and a way more kludgy mechanism IMO) alternative that I see would be to make some new property in **ESAPI.properties**, such as
```
    Encoder.LDAP.isActiveDirectory=false
```
We'd have it default to `false` so that it wouldn't change existing behavior.

@xeno6696 and @jeremiahjstacey - What are your thoughts on this?

Reply:
> [RFC 2254](https://datatracker.ietf.org/doc/html/rfc2254), and if you look at Section 4 of that document (which defines search filters for LDAPv3), you will not find '/' as being mentioned as requiring encoding. So, I am thinking this is something that is Active Directory specific. (Thanks Microsoft! Another standard ignored.)

I was thinking... there are three well known X.500 directories: Microsoft's Active Directory, Novell's eDirectory, and U&L LDAP which is based on the RFCs. Microsoft requires forward slash escaping, but the other two may consume an escaped forward slash.

It looks like [RFC 4517](https://datatracker.ietf.org/doc/html/rfc4517) specifies an escaped forward slash in its grammar:

```
PrintableCharacter = ALPHA / DIGIT / SQUOTE / LPAREN / RPAREN /
                           PLUS / COMMA / HYPHEN / DOT / EQUALS /
                           SLASH / COLON / QUESTION / SPACE
      PrintableString    = 1*PrintableCharacter
      IA5String          = *(%x00-7F)
      SLASH              = %x2F  ; forward slash ("/")
      COLON              = %x3A  ; colon (":")
      QUESTION           = %x3F  ; question mark ("?")
```
I'm wondering if the other Directories will just accept the encoded version of the forward slash so you don't need `boolean isActiveDirectory`.

Reply:
I asked if OpenLDAP handles the escaped forward slash as expected on the [openldap-technical mailing list](https://lists.openldap.org/hyperkitty/list/openldap-technical@openldap.org/2022/6/). The message is being held for moderation, but it should post soon.

Reply:
@kwwall I would say that this is what the Codecs were designed for.  Fork the LDAP Codec, and create one for each desired encoding/decoding instance.  Implementers that have to use specific LDAP codecs would then have that choice and could set them in `ESAPI.properties`

Unless I'm missing something this is a quick fix.